src/protocolsupport/protocol/packet/middleimpl/clientbound/play/v_pe/DeclareCommands.java

```diff
@@ -0,0 +1,473 @@
+			if ((flag & FLAG_HAS_MIN_VALUE) != 0) {
+				int min = from.readInt();
+			}
+			if ((flag & FLAG_HAS_MAX_VALUE) != 0) {
+				int max = from.readInt();
+			}
+		});
+		argTypeReaders.put("brigadier:float", from -> {
+			byte flag = from.readByte();
+			if ((flag & FLAG_HAS_MIN_VALUE) != 0) {
+				float min = from.readFloat();
+			}
+			if ((flag & FLAG_HAS_MAX_VALUE) != 0) {
+				float max = from.readFloat();
+			}
+		});
+		argTypeReaders.put("brigadier:double", from -> {
+			byte flag = from.readByte();
+			if ((flag & FLAG_HAS_MIN_VALUE) != 0) {
+				double min = from.readDouble();
+			}
+			if ((flag & FLAG_HAS_MAX_VALUE) != 0) {
+				double max = from.readDouble();
+			}
+		});
+		argTypeReaders.put("minecraft:entity", from -> {
+			// The flag determines the amount (single or double) and type (players or entities)
+			// See FLAG_ENTITY_AMOUNT_IS_SINGLE and FLAG_ENTITY_TYPE_IS_PLAYER.
+			byte flag = from.readByte();
+		});
+		argTypeReaders.put("minecraft:score_holder", from -> {
+			// The "multiple" boolean is true if multiple, false if single.
+			boolean multiple = from.readBoolean();
+		});
+	}
+
+	private static String readArgType(ByteBuf from) {
+		String argType = StringSerializer.readVarIntUTF8String(from);
+
+		// Depending on argType, there might be additional data.
+		// At this point, we're just throwing this away, but we need at least
+		// skip over it in the buffer.
+		// If we have no registered remap, there is no additional data.
+		// Note that this might need updating in future versions of Minecraft.
+		argTypeReaders.getOrDefault(argType, o -> {}).accept(from);
+
+		return argType;
+	}
+
+	@Override
+	public RecyclableCollection<ClientBoundPacketData> toData() {
+		PECommandsStructure struct = buildPEStructure();
+
+		return RecyclableSingletonList.create(createPacket(struct));
+	}
+
+	private PECommandsStructure buildPEStructure() {
+		PECommandsStructure peStruct = new PECommandsStructure();
+
+		// For every command, we need to create a set of overloads, i.e. the different
+		// ways the command can be called, with a list of argument nodes for each alternative.
+		for (CommandNode node : topLevelNodes) {
+			List<List<PEArgumentNode>> overloads = new ArrayList<>();
+			walkNode(peStruct, overloads, node, new ArrayList<>());
+			peStruct.registerOverloads(node, overloads);
+		}
+
+		return peStruct;
+	}
+
+	private void walkNode(PECommandsStructure peStruct, List<List<PEArgumentNode>> overloads, CommandNode currentNode, List<CommandNode> nodePath) {
+		if (currentNode.isLeaf()) {
+			List<PEArgumentNode> argumentList = new ArrayList<>();
+			for (CommandNode node : nodePath) {
+				PEArgumentNode peNode = new PEArgumentNode(node, peStruct);
+				argumentList.add(peNode);
+			}
+			overloads.add(argumentList);
+		} else {
+			for (int childNodeIndex : currentNode.getChildren()) {
+				CommandNode childNode = allNodes[childNodeIndex];
+				List<CommandNode> newPath = new ArrayList<>(nodePath);
```

Comment:
Might be worth investigating if clone can be avoided.
For example you may add a childnode and then remove it after calling walkNode.

Reply:
Maybe, and maybe not. I think there's a limit to how much effort I'm willing to spend on micro-optimizations on this sent-once-per-login packet...